Update Customer.io destination to use v2 api

nunofgs opened the pull request:
Updates the Customer.io destination:

* Uses the [v2 API](https://www.customer.io/docs/api/track/#tag/track_v2) whenever possible.
* Adds batching to all supported actions.
* Adds new actions:
  - deleteRelationship
  - deleteObject
  - deletePerson
  - mergePeople
  - suppressPerson
  - unsuppressPerson
  - reportDeliveryEvent
* Adds `relationship_attributes` to the _createUpdateObject_ and _createUpdatePerson_ actions.
* Fixes the default `object_type_id` field on the _createUpdateObject_ and _createUpdatePerson_ actions.
* Updates the default subscription of `trackEvent` to ensure semantic events are excluded.
* Changes the "site ID" and "api key" fields to be of type password.
* Ensures all tests exercise available regions (US + EU) as well as batch/non-batch variants.
* Fixes sharing of `$.traits` field in the _createUpdatePerson_ action (by changing the default for new users).
* Removes `Delete Device` from presets to avoid misusing the action.

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

nunofgs commented on packages/destination-actions/src/destinations/customerio/__tests__/deleteDevice.ts:
```diff
@@ -1,113 +0,0 @@
-import nock from 'nock'
```
This file was renamed. It was missing the `.test.ts` prefix.

joe-ayoub-segment commented:
Thanks for the PR @nunofgs . I'll schedule time to review.

joe-ayoub-segment commented:
hi @nunofgs - apolgies for not getting to this sooner. 
I'll take a look now. 

This is a large change, so I'd like to request that you provide some sort of testing plan if possible. The 2 reasons for this are so that: 
1. Customers who are already using the Integration won't be negatively impacted. i.e pre-existing functionality continues to work.  
2. The new functionality works as expected.  

Would you be available to catch up face to face so we can review together?
Here is my Calendly link: http://calendly.com/joe_ayoub/integration-workshop 

Best regards,
joe

nunofgs commented:
Hi Joe, thanks for taking a look. Of course! That's very sensible.

I've set up some time to discuss tomorrow. See you then!

joe-ayoub-segment commented on packages/destination-actions/src/destinations/customerio/utils.ts:
```diff
@@ -1,17 +1,11 @@
-export enum AccountRegion {
-  US = 'US 🇺🇸',
-  EU = 'EU 🇪🇺'
+const isEmail = (value: string): boolean => {
+  return !!validateEmail(value)
 }
```
Hi @nunofgs , 

When I tried deploying the code the error message I saw indicated that there was a problem with AJV not knowing how to handle email. I think that this is because it wasn't aware of the 'email' format.

Here's the error message from our build pipeline: 

```
Error: unknown format "email" ignored in schema at path "#"
--
  |  
  | › unknownFormat (node_modules/ajv/lib/vocabularies/format/format.ts:87:15)
  | › validateFormat (node_modules/ajv/lib/vocabularies/format/format.ts:75:9)
  | › Object.code (node_modules/ajv/lib/vocabularies/format/format.ts:40:10)
  | › keywordCode (node_modules/ajv/lib/compile/validate/index.ts:532:9)
  | › node_modules/ajv/lib/compile/validate/index.ts:265:9
  | › CodeGen.code (node_modules/ajv/lib/compile/codegen/index.ts:525:33)
  | › CodeGen.block (node_modules/ajv/lib/compile/codegen/index.ts:680:20)
  | › iterateKeywords (node_modules/ajv/lib/compile/validate/index.ts:262:7)
  | › groupKeywords (node_modules/ajv/lib/compile/validate/index.ts:241:7)
  | › node_modules/ajv/lib/compile/validate/index.ts:233:38

```

I looked into this and the following code might be worth trying as it's already in use with another integration: 

```
import { fullFormats } from 'ajv-formats/dist/formats'

  if (!(fullFormats.email as RegExp).test(email)) {
    throw new PayloadValidationError("Email provided doesn't seem to be in a valid format.")
  } 
```

You can see this code in use here. https://github.com/segmentio/action-destinations/blob/main/packages/destination-actions/src/destinations/google-enhanced-conversions/functions.ts#L163

nunofgs commented on packages/destination-actions/src/destinations/customerio/createUpdateObject/index.ts:
```diff
@@ -35,6 +41,15 @@ const action: ActionDefinition<Settings, Payload> = {
+        'Optional attributes for the relationship between the object and the user. When updating an relationship, attributes are added or updated, not removed.',
+      type: 'object',
+      default: {
+        '@path': '$.context.relationshipTraits'
```
Updated to traits, [per the group spec](https://segment.com/docs/connections/spec/group/#:~:text=for%20more%20detail.-,traits,-optional).

nunofgs commented on packages/destination-actions/src/destinations/customerio/mergePeople/index.ts:
```diff
@@ -0,0 +1,60 @@
+  fields: {
+    primary: {
+      label: 'Primary User',
+      description: `The person that you want to remain after the merge, identified by id, email or cio_id. This person receives information from the secondary person in the merge.`,
```
No, it can only be used to merge 2 previously-identified users.

nunofgs commented on packages/destination-actions/src/destinations/customerio/mergePeople/index.ts:
```diff
@@ -0,0 +1,60 @@
+
+function mapPayload(payload: Payload) {
+  return {
+    primary: resolveIdentifiers({ person_id: payload.primary }),
+    secondary: resolveIdentifiers({ person_id: payload.secondary })
```
The internal ID we assign to each profile will always start with `cio_`. However, the user can also provide their own ID.

More about our supported identifiers [can be found here](https://customer.io/docs/api/track/#operation/track:~:text=You%20may%20identify%20a%20person%20by%20id%2C%20email%20address%2C%20or%20the%20cio_id%20(when%20updating%20people)%2C%20depending%20on%20your%20workspace%20settings.).

nunofgs commented on packages/destination-actions/src/destinations/customerio/createUpdateDevice/index.ts:
```diff
@@ -59,24 +67,29 @@ const action: ActionDefinition<Settings, Payload> = {
-      lastUsed = convertValidTimestamp(lastUsed)
-    }
+  perform: (request, { payload, settings }) => {
+    return sendSingle(request, { action: 'add_device', payload: mapPayload(payload), settings, type: 'person' })
```
It upserts, yes.

nunofgs commented on packages/destination-actions/src/destinations/customerio/createUpdateObject/index.ts:
```diff
@@ -23,7 +25,7 @@ const action: ActionDefinition<Settings, Payload> = {
       type: 'string',
       default: {
-        '@template': '{{traits.created_at}}'
+        '@path': '$.timestamp'
```
I changed both of them to $.timestamp after our conversation, but then reverted the `createUpdatePerson` (the only other created_at in our destination). This is because an upsert would cause the person's "created_at" to be updated on every call. I'll revert this one too, for the same reason.

nunofgs commented on packages/destination-actions/src/destinations/customerio/createUpdateObject/index.ts:
```diff
@@ -70,49 +84,100 @@ const action: ActionDefinition<Settings, Payload> = {
+    id,
+    convert_timestamp,
+    custom_attributes,
+    relationship_attributes,
```
Do you see any issues with it at the moment? We're already handling it conditionally in the mapPayload() function.

nunofgs commented on packages/destination-actions/src/destinations/customerio/createUpdateObject/index.ts:
```diff
@@ -70,49 +84,100 @@ const action: ActionDefinition<Settings, Payload> = {
+    object_id: id
+  }
+
+  let rel_attrs = relationship_attributes as Record<string, unknown>
```
Do you see any issues with it at the moment? We're already handling it conditionally in the mapPayload() function.

nunofgs replied:
I added a regression test to ensure BC.

nunofgs commented on packages/destination-actions/src/destinations/customerio/createUpdateDevice/index.ts:
```diff
@@ -59,24 +67,29 @@ const action: ActionDefinition<Settings, Payload> = {
-      }
-    })
+function mapPayload(payload: Payload) {
+  const { device_id, platform, last_used, attributes, ...rest } = payload
```
I believe it already is. I added a regression test to ensure this.

nunofgs commented on packages/destination-actions/src/destinations/customerio/createUpdatePerson/index.ts:
```diff
@@ -77,57 +89,63 @@ const action: ActionDefinition<Settings, Payload> = {
-      }
-    }
+function mapPayload(payload: Payload) {
+  const { id, custom_attributes = {}, relationship_attributes, created_at, group_id, object_type_id, ...rest } = payload
```
I believe it already is. I added a regression test to ensure this.

nunofgs commented on packages/destination-actions/src/destinations/customerio/utils.ts:
```diff
@@ -78,3 +86,123 @@ export const convertAttributeTimestamps = (payload: Record<string, unknown>): Re
+    return
+  }
+
+  const [{ settings }] = options
```
Since this is just an internal helper function, it just made things a bit cleaner and more consistent with the single `perform` function, like this:

```
  performBatch: (request, { payload: payloads, settings }) => {
    return sendBatch(
      request,
      payloads.map((payload) => ({ action: 'identify', payload: mapPayload(payload), settings, type: 'person' }))
    )
  },

  perform: (request, { payload, settings }) => {
    return sendSingle(request, { action: 'identify', payload: mapPayload(payload), settings, type: 'person' })
  }
```

A separate parameter will become a bit more verbose:

```
  performBatch: (request, { payload: payloads, settings }) => {
    return sendBatch(
      request,
      {
        payloads: payloads.map((payload) => ({ action: 'identify', payload: mapPayload(payload), type: 'person' }))
        settings
      }
    )
  },
```

Happy to do it if you feel strongly though!

joe-ayoub-segment commented on packages/destination-actions/src/destinations/customerio/deleteObject/index.ts:
```diff
@@ -0,0 +1,55 @@
+const action: ActionDefinition<Settings, Payload> = {
+  title: 'Delete Object',
+  description: 'Delete an object in Customer.io.',
+  defaultSubscription: 'event = "Group Deleted"',
```
Will the term 'Group Deleted' make sense to a customer? Or would something like 'Object Deleted' be better?

nunofgs replied:
Absolutely. Nice catch.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/customerio/suppressPerson/index.ts:
```diff
@@ -0,0 +1,49 @@
+    person_id: {
+      label: 'Person ID',
+      description:
+        'The ID used to uniquely identify a person in Customer.io. Any valid identifier is supported, such as ID, cio_id, email, etc. Anonymous users cannot be suppressed. [Learn more](https://customer.io/docs/identifying-people/#identifiers).',
```
hi @nunofgs how will customer.io know if the user passes in a cio_id or an email or an ID? Is there some code which checks this? I don't see such a check happening in the perform() function below...

nunofgs replied:
That's done in the `resolveIdentifiers` function [here](https://github.com/segmentio/action-destinations/pull/1800/files#diff-988097df49c2cdf17bc2ccdad58f1c80a953763439c7660af552df0d9b356562R157).

joe-ayoub-segment commented:
hi @nunofgs - how has testing gone? Do you think we can deploy?

nunofgs commented:
Hi! Yes, please. I’m all set on my side.

joe-ayoub-segment commented:
Thanks @nunofgs - would it be possible to add some proof of testing for the old and new Actions please?
I'm about to run a deploy today (in the next hour) so would love to get this out if possible!

joe-ayoub-segment commented:
Hi @nunofgs I ran CI and there's a few failures.

joe-ayoub-segment commented:
The problem is here: 

$ yarn tsc -b tsconfig.build.json
$ /home/runner/work/action-destinations/action-destinations/node_modules/.bin/tsc -b tsconfig.build.json
Error: src/destinations/customerio/mergePeople/index.ts(34,7): error TS2345: Argument of type '{ action: string; payload: { primary: Identifiers | undefined; secondary: Identifiers | undefined; }; settings: Settings; type: string; }[]' is not assignable to parameter of type 'RequestPayload<BasePayload>[]'.
  Type '{ action: string; payload: { primary: Identifiers | undefined; secondary: Identifiers | undefined; }; settings: Settings; type: string; }' is not assignable to type 'RequestPayload<BasePayload>'.
    Types of property 'payload' are incompatible.
      Type '{ primary: Identifiers | undefined; secondary: Identifiers | undefined; }' has no properties in common with type 'BasePayload'.
Error: src/destinations/customerio/mergePeople/index.ts(46,7): error TS2559: Type '{ primary: Identifiers | undefined; secondary: Identifiers | undefined; }' has no properties in common with type 'BasePayload'.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

<img width="746" alt="image" src="https://github.com/segmentio/action-destinations/assets/45374896/3bef28ee-5db1-4ee2-9021-254155d192e1">


I'm not sure what the structure of your payload should be so I can't figure out a fix.

joe-ayoub-segment commented:
hi again @nunofgs - apologies for all the messaging! 

I removed the Action which is causing CI to fail and deployed the code to Staging. 

Would you be available to jump on a call so we can do some end to end testing please?

Here's my calendly: http://calendly.com/joe_ayoub/integration-workshop 

Best regards
Joe

joe-ayoub-segment commented:
Lots of feedback on this PR today!

I've noticed that the default mappings for person_id field across different Actions seem to vary. 

Some of them have the following: 

     ```
description:
        'The ID used to uniquely identify a person in Customer.io. Any valid identifier is supported, such as ID, cio_id, email, etc. Anonymous users cannot be unsuppressed. [Learn more](https://customer.io/docs/identifying-people/#identifiers).',
 default: {
        '@if': {
          exists: { '@path': '$.userId' },
          then: { '@path': '$.userId' },
          else: { '@path': '$.traits.email' }
        }
      },
```

And other have this: 

  ```
 description:
        'The ID used to uniquely identify a person in Customer.io. [Learn more](https://customer.io/docs/identifying-people/#identifiers).',
    default: {
        '@path': '$.userId'
      }
```

Should the descriptions and default mappings be the same across all Actions?

If you do decide to keep email in the default path for track() based Actions, then make sure it's pointing to $.context.traits.email and not $.traits.email . 

If you want to have email in the default path for identify() based Actions, then it should point to $.traits.email .

nunofgs commented:
Thanks Joe. I've scheduled some time for us.

In the meantime I fixed the type error and updated the descriptions and defaults for person_id. Also, here's some screenshots of the tests:

<img width="1570" alt="Screenshot 2024-02-14 at 23 16 44" src="https://github.com/segmentio/action-destinations/assets/92085/8a7df5b3-aa52-43c9-9c09-636b17614efc">
<img width="1234" alt="Screenshot 2024-02-14 at 23 17 01" src="https://github.com/segmentio/action-destinations/assets/92085/ef75ed61-6520-40b0-9a24-1682701e4602">

I can provide screenshots for every change but might be a bit much to post them here. Want them in any particular format?

joe-ayoub-segment commented:
> I can provide screenshots for every change but might be a bit much to post them here.

Thanks Nuno. 
If you could post a couple more here that's OK. If you say that you've tested all the Actions / Presets I believe you!

joe-ayoub-segment commented:
BTW @nunofgs a few files outside of the customerio folder were edited. It's only formatting changes so should be safe. If you have a minute maybe they could be removed from the PR?

nunofgs commented:
> BTW @nunofgs a few files outside of the customerio folder were edited. It's only formatting changes so should be safe. If you have a minute maybe they could be removed from the PR?

Done.

joe-ayoub-segment commented:
Hi @nunofgs are you confident that the changes are safe / fully tested?
I've labeled this for release so it will go out tomorrow.

nunofgs commented:
Hi Joe. I am confident in the changes but I'll run through everything again before our meeting tomorrow. See you then!

nunofgs commented:
A few more screenshots, as requested:

<img width="2323" alt="deleteObject" src="https://github.com/segmentio/action-destinations/assets/92085/dd84c6e3-09b2-424c-871f-ad14500b19e6">
<img width="2332" alt="deletePerson" src="https://github.com/segmentio/action-destinations/assets/92085/621f4a79-d374-40ab-a8d9-dea588b7b4cf">
<img width="2357" alt="suppress" src="https://github.com/segmentio/action-destinations/assets/92085/218a357e-618e-4a89-a9cc-fa1c0e41e06f">
<img width="2304" alt="unsuppress" src="https://github.com/segmentio/action-destinations/assets/92085/810caa87-c274-49a1-95fc-a0e34de0da4a">
<img width="2315" alt="mergePeople" src="https://github.com/segmentio/action-destinations/assets/92085/ffb70d3f-ae0b-4e2d-8018-16e333965063">
<img width="2299" alt="trackScreenView" src="https://github.com/segmentio/action-destinations/assets/92085/ef99d412-b7c0-4d1a-8c5c-1cccb49bab03">
<img width="2301" alt="createUpdatePerson with relationshipAttributes" src="https://github.com/segmentio/action-destinations/assets/92085/c5fc9fac-ac9a-44e0-8253-db36bfb643d5">
<img width="2300" alt="deleteRelationship" src="https://github.com/segmentio/action-destinations/assets/92085/aa2f098c-8f4e-45a6-adc2-5782df22226c">

joe-ayoub-segment commented:
Hi @nunofgs this PR has been deployed!
